### DISCUSSION TITLE
Design for "Remove From Continue Watching"

### DISCUSSION THREAD

**Discussion by OccamsChainsaw42** [jellyfin/jellyfin-meta#50] [Ideas]
Here is the write up of the current design we have for implementing this feature.

1. Add an entry to the item context menu for "Remove from Continue Watching"
2. Create a new boolean flag "ExcludedFromContinueWatching"
3. The command in the context menu will set the flag to true and save it to the database
4. Modify the query that sources Continue Watching to exclude entries with this flag.
5. When a user starts playing an object set the flag to false so the item can show again in Continue Watching

This seems pretty straight forward.

We have looked into how to implement this feature and have a few concerns:

- The flag needs to be store per user similar to lastPlayedDate, playbackPosition etc. 
- So we would likely need a new column in the "UserDatas" table. 
- Would this require the creation of a migration script?

It seems we would need to modify the "InternalItemsQuery" object to be able to include this flag.
I assume we would also need to modify the SqliteItemReposity class to modify the generation of the query.

An alternative to modifying InteralItemsQuery would be to separately get a list of guids that have the exclude flag for the current user
and include them in the "ExcludeItemIds" property of the query. This might be more straight forward if there is a way to do a more manual sql query to get the list of guids?

An alternative to the flag design as a whole would be to instead offer the user the option to reset the watch time. 
This would remove the need to make any modifications to the database or query generation.
The advantage of the flag would be to keep the users current watch position but still exclude it.

**Comment by nathangur** [comment]
This path is a bit different from my implementation. I've gone down the route of simply removing the item from Continue Watching without resetting the user's position in the piece of media.

**Reply by nathangur** [reply]
I'm currently struggling with making the connection between the server and the web client as I'm getting some api errors.

**Reply by OccamsChainsaw42** [reply]
So the main method we've proposed of using a flag wouldn't affect the users position. Resetting the users position is an alternative method of removing it from continue watching since only items with some play time will be considered

**Reply by nathangur** [reply]
I see. If you are fine with it, I'd like to help you guy's with continuing your implementation. I don't think I can get mine to work in it's current state without some major changes

**Reply by BrennanWoodbury** [reply]
Sure thing, DM'd you on element.

**Comment by neopc10** [comment]
I'd love to see this extended to "Remove from Next Up" if there's an appetite for that.

**Reply by OccamsChainsaw42** [reply]
We went ahead and added this feature as well in a separate PR.

**Comment by nathangur** [comment]
@cvium Why did you reopen this discussion? Was there something you'd like to add?

**Reply by thornbill** [reply]
As I commented on one of the open PRs related to this:

> Sorry if the process hasn't been clear. The Jellyfin team needs to be involved in the discussion to make sure the implementation aligns with best practices, etc. That is the main reason a discussion is needed first for changes of a larger scope or fundamental nature.

**Reply by nathangur** [reply]
Ah I see!

**Comment by JPVenson** [comment]
What is the benefit of this over just marking the item as played via the check icon?

**Reply by nathangur** [reply]
Doing that causes it to be marked as played, meaning it won't show up in the next up category. This has been highly requested on the jellyfin requests page for people that are admins and others that don't want their media to be marked as played. I myself find this especially annoying for shows where when rewatching it, none of them appear in the next up category.

**Reply by JPVenson** [reply]
Then i would be in favor to set the watch state back and remove the watch flag instead.
This also returns the media item back to its "pre-watched" state and puts it back into the NextUp queue.
That is more or less also what i am usually doing when rewatching a whole series, i mark everything as watched and then click on the button to unmark it as watched. That resets the state just fine.

So marking it as watched and then unwatched does the same without introducing a new DB field(s) and more. We could maybe, if not already there, create a new Api method and button in the UI to "unwatch" or reset the watch state instead.

**Reply by thornbill** [reply]
This seems like a good idea to me. The issue with marking as watched then unwatched is that it still updates the watch count... so a dedicated way to reset the watch status seems worthwhile.

**Reply by JPVenson** [reply]
Then i would go the full route and introduce 3 methods:
- Unwatch Episode
- Unwatch Season
- Unwatch Series
(or depending on the media item)
all resetting the watch status back to unwatched and reset the progress.

**Reply by thornbill** [reply]
Now I'm second guessing if just resetting the watch status is "good enough". I tested some streaming services and they have the ability to remove items without losing the watch status. 🤷‍♂️

**Reply by JPVenson** [reply]
Well i would prefer not to build that much logic into the current system but maybe rethink it. The reset of the watch status would be easy to implement, get the job done and does not introduce more code then absolutely necessary.

**Reply by nathangur** [reply]
Another reason why we added a DB field is for having a timer to automatically remove items from continue watching after a certain period of time without changing the watch status. This is pretty highly requested, and people like to keep things clean without wiping where they left off. And, when you click the watched button, it adds it to next up, cluttering that when you are really just trying to clean up your library. 

Little things like this that are just assumed to be good enough for the user just adds a little extra edge to the Jellyfin experience. At least in my opinion :)

**Reply by nathangur** [reply]
Back again to talk about this implementation. I've had a bunch of items that I want to exclude from continue watching or next up and have had to mark the whole thing as watched. It's genuinely quite annoying to deal with.

**Reply by JPVenson** [reply]
Well the server team discussed the current PR and the general idea. We like the idea and are overall very dissatisfied how the current next-up system works. We decided that any change to the current system just adds code to a complex system that needs a complete overhaul anyway. The idea is worthwhile but we dont want to change the existing code for now.

**Comment by spooksbit** [comment]
>An alternative to the flag design as a whole would be to instead offer the user the option to reset the watch time.

This appears to be achievable right now by making a POST request to `/Users/{userId}/Items/{itemId}/UserData` with `{"PlaybackPositionTicks": 0}`. From my own testing, I am finding that items are being removed after making the request, but other fields appear to be retained (ie last played, play count, etc).

Would this be an acceptable solution?

**Reply by nathangur** [reply]
Yes it does work, but it doesn't accomplish. What removed from continue watching is supposed to do. It's supposed to retain the playback position but remove the item from continue watching or the next up category. Currently the only way to accomplish this is to mark the item as watched which is not ideal.

**Reply by spooksbit** [reply]
My apologies if this has already been discussed, but what is the value proposition in retaining playback position after removing items from this list? This action seems to carry an explicit intent to forget where one has stopped playback, rather than an intent to retain playback position. 

Comments from the feature request do not indicate this as a requirement either. To the contrary. resetting the playback position has been noted as "acceptable" behavior by some users and even desired by others. Maybe I'm the minority here, but it feels more like an anti-feature to retain playback position. 

Is there maybe a different feature that is being implemented by requiring this behavior?

**Reply by nathangur** [reply]
> My apologies if this has already been discussed, but what is the value proposition in retaining playback position after removing items from this list? This action seems to carry an explicit intent to forget where one has stopped playback, rather than an intent to retain playback position. 

> Comments from the feature request do not indicate this as a requirement either. To the contrary. resetting the playback position has been noted as "acceptable" behavior by some users and even desired by others. Maybe I'm the minority here, but it feels more like an anti-feature to retain playback position. 

Actually I think the opposite is true. Currently you can mark an item as watched and that resets the playback position and marks it as watched. Then you can just uncheck it again to have it completely reset. However, the feature request wanted to retain the watch position but remove the item from continue watching or next up. This is especially useful for admins who play media for testing and have a lot of items on their front page, and if the PR's relating to this discussion had been accepted then a timer setting would have been proposed to remove items from either category after a period of time, similar to a feature Plex has. 

> Is there maybe a different feature that is being implemented by requiring this behavior?

The only thing that was planned was the timer. I personally would have used this (and am currently using it on my dev server) to keep things clean and uncluttered. Which is essentially what this feature aimed to accomplish, cleaning up the list of 100+ items from my continue watching category but not erasing the playback position. Is this a necessary feature? By all means no, but I am a firm believer that the little things that may seem insignificant or not useful actually add to the user experience. Even if a user never uses it, having the ability to do something solidifies the user experience.

And thus concludes my long winded, unneeded essay :)

**Reply by spooksbit** [reply]
>However, the feature request wanted to retain the watch position but remove the item from continue watching or next up

Here are some comment excerpts from the feature request which seem to contradict this sentiment:

>It would definitely be fine for the watched progress to be reset when it's removed.

>If it can help or give ideas, [X] offers an option on items that have been partly watched to "reset resume time". I use it fairly often and it is a nice way to avoid the problem around changing the item's watched status

>The "Continue Watching" just gets anoying and in the way if you cant remove stuff. I also like [X]'s option to reset resume time.

At the very least, it sounds like an option to reset playback position could be submitted as a feature request in the future.

In the case of testing media, wouldn't the retained playback position be the time that was chosen for testing purposes? If it is (and please correct me if I'm wrong), I would think it's desirable to discard that position instead of retaining it. If I was testing a movie and skipped to the middle to validate subtitles worked and removed it from the list, I would not want to be returned to that spot when I go to actually watch it later. From the comments, it sounds like usually the first few minutes of an item is used for testing which is not significantly different from the playback position of 0. If anything, I could see the option of resetting playback to 0 being presented as an enhancement by users to better support that specific use case since otherwise each item would play from some arbitrary point near the start and it would make sense to just reset the position.

Maybe I am the minority here, but retaining playback position sounds more like an annoying side effect and I feel would take away from user experience rather than add to it. It also seems that the proposed changes are at an impasse for legitimate concerns, but it also seems like those concerns are based on an implementation that is unnecessarily too strict. Instead, I think resetting the playback position relaxes the implementation in a way that not only eliminates the maintenance concerns but is also more desirable by users. Win-Win.

**Reply by nathangur** [reply]
I think you misunderstood what this implementation is. This is for an extra button in the action menu for an item. It's not replacing marking an item as watched.

**Reply by spooksbit** [reply]
I would not expect it to replace marking an item as watched? Clearing the position does not mark an item as watched so that button would continue to exist.

See: https://github.com/spooksbit/jellyfin-web/commit/9cc592d2971d9ffa6240a8b681031c9fdbcd7b07

**Comment by jcreek** [comment]
Hi everyone, I'm a software engineer who's just moving to Jellyfin after all the recent Plex changes. I'm looking to get involved in contributing, and this issue seemed a good one to look at seeing as there's high community demand and ongoing discussion for the last few years on the best way to implement it to minimise technical debt and unnecessary complexity.

I'm summarising some key information below, please do correct me if I've misunderstood or misrepresented something:

* @westonhowe98’s original design introduced a per-user `ExcludedFromContinueWatching` flag in `UserDatas` to hide an item while preserving playback position.
* Several people noted the extra column/migration and the need to touch `InternalItemsQuery`/`SqliteItemRepository`, raising worries about adding complexity to an already-hairy part of the codebase.
* The server team summarised the hesitation clearly: any tweak to the current Resume / Next-Up pipeline “just adds code to a complex system that needs a complete overhaul anyway.”
* Others suggested simply resetting `PlaybackPositionTicks` to 0 via the existing user-data API as a quick-win, although opinions differ on whether losing the resume point is desirable. 

I'd like to propose a middle-ground approach, where we store CW as a separate table. Instead of a suppression flag or wiping the resume point, Continue Watching could be an explicit per-user list:

| Event | Action | Effect on CW |
|-------|--------|--------------|
| Playback passes *MinResume%* | `UPSERT (UserId, ItemId)` in `UserContinueWatching` | shows/bumps item |
| User hits “Remove from CW” | `DELETE` the row | disappears |
| Playback finishes *(≥ MaxResume% or mark watched)* | `DELETE` row | disappears |
| User starts item again | row auto-re-inserted | re-appears |

This would address several of the concerns I've seen:
* Zero changes to `UserDatas` – this avoids the migration and wide-query edits flagged above.  ￼
* No extra logic in `InternalItemsQuery` – `ItemsController.GetResumeItems` would become a simple `SELECT` from the new table, leaving the existing (admittedly gnarly) resume/next-up queries untouched.
* Keeps resume data intact for users who value it (per @nathangur’s point about admins/testing) while letting others still wipe progress by using the mark watched button if they prefer.  ￼
* Aligns with the planned overhaul – moving CW into its own store is a clean first step toward decoupling all recommendation lists, something the team already wants to tackle, if I've understood the discussions correctly.

I anticipate this would have minimal impact on the server-side. 
* Playback hooks (already fire in PlaybackManager) just call UpsertCW / RemoveCW.
*A single new route `POST /Users/{userId}/ContinueWatching/{itemId}/Remove` deletes the row.
* Clients need only add one context-menu option; nothing else changes.

I’m happy to prototype this and open a draft PR so the team can see the diff size in real terms.

Does this feel like a worthwhile compromise, or are there blockers I’m overlooking? I'm also conscious that I've spent very little time in this codebase so I may be mis-interpreting or mis-remembering some of the current architecture or functionality. 

Thanks for considering!

**Reply by barrelltitor** [reply]
I think a core thing of the disagreement there is 
```
We like the idea and are overall very dissatisfied how the current next-up system works.
```
They want to overhaul the next-up system too. Maybe if you're happy to take on both there's a higher chance of getting this through? Maybe the next up system could have a similar separate table as I can see myself wanting to remove things from next up as well

E.g maybe I watched one episode of something and I don't want to "unwatch" it, but at the same time I don't want it showing next up either, so this is basically the same scenario. You'd hit 2 birds with one stone

**Reply by OccamsChainsaw42** [reply]
I believe another concern was that the project was in the process of migrating to an ORM instead of using direct SQL, and the way I originally implemented this feature added more raw SQL calls, which contributed to the problem.

I'm not sure if the ORM migration has been fully completed, but I know it's much further along now. It would probably be much easier, and cleaner, to implement this feature using the ORM at this point.

**Reply by JPVenson** [reply]
the rewrite to EFCore (ORM) is completed and will be shipped soon with 10.11
But you are right as our aim is to simplify and also diversify our database backend in the very near future, Raw SQL is a absolute No-Go.

**Reply by ChristopheAwad** [reply]
Hi, I appreciate everyone's work on this project and this feature specifically. Wanted to ask if there is any progress on this, what the next steps are?

**Reply by nathangur** [reply]
> Hi, I appreciate everyone's work on this project and this feature specifically. Wanted to ask if there is any progress on this, what the next steps are? 

Hey there. I'll look back into this soon since 10.11 is near stable and the EFCore backend is now in place. I've never worked with it before so no promises

**Reply by BrennanWoodbury** [reply]
@nathangur I have plenty of experience with efcore, are you still digging into this, happy to help where I can if it'd be helpful.

**Reply by nathangur** [reply]
@BrennanWoodbury Thanks for reminding me. I'm pulling the latest and I'm going to re-familiarize myself with the database.

**Reply by cvium** [reply]
The design seems reasonable to me. There's a couple of things that aren't explicitly described though:

1. How are newly added episodes handled? And deletions?
2. What happens if an episode is removed from CW and there's a next one? Is it the whole series that is removed or does it skip to the next episode?

**Reply by Soitora** [reply]
> The design seems reasonable to me. There's a couple of things that aren't explicitly described though:
> 
> 
> 
> 1. How are newly added episodes handled? And deletions?
> 
> 2. What happens if an episode is removed from CW and there's a next one? Is it the whole series that is removed or does it skip to the next episode?

If a series is manually Removed from CW then it should stay that way until a user has watched an episode again to remove the flag. You're removing the entire series and not just an episode

New episodes or deletions shouldn't affect anything

**Reply by cvium** [reply]
That depends what you put in the table. Is the new table a snapshot of CW items? If so, then additions and deletions should affect it. If a series is removed (not sure how that would be tracked), nothing happens if an episode is added, but otherwise it should be put into CW. Likewise for episode deletions.

If the CW table is tracked on a series level (and movies, but they are simple), then you still need a very awkward query to determine which is the "next" episode based on the user's history. I don't like this approach though cause we currently have a very awkward query.

**Reply by nathangur** [reply]
@cvium @Soitora To answer your concerns:

The table doesn't hold the continue watching list, it just holds what you've removed from it. The resume and next up queries are untouched, they only skip anything with a row in there. And new episodes aren't really a special case because there's nothing hiding them so they just show up like normal. And if an item gets deleted the row goes with it as a FK cascade.

For the series question, I separated it into two actions rather than one. Remove from continue watching hides just that episode and remove from next up hides the whole series from next up. Originally I had it collapsing onto the series for CW and it became
genuinely annoying in practice where older episodes that were just about done (24/25min) that I would want to remove also removed my current in-progress episode. Splitting it also means you never have to work out what the "next" episode is, I just filter the input to the existing next up query.

**Reply by Soitora** [reply]
@nathangur Yeah, that makes sense. Since Continue Watching is essentially for resuming a specific episode, removing only that episode feels like the right behavior. “Remove from Next Up,” on the other hand, would suppress the entire series from being suggested.

I was initially thinking about it from a Plex perspective, where both behaviors are presented within a single Continue Watching section. Jellyfin separates Continue Watching and Next Up, so having two distinct actions fits its existing structure well.

Sounds like a good approach to me.

**Reply by nathangur** [reply]
I implemented this for the server and web repos and both commits can be found here:

https://github.com/jellyfin/jellyfin/compare/master...nathangur:jellyfin:RemoveFromContinueWatching
https://github.com/jellyfin/jellyfin-web/compare/master...nathangur:jellyfin-web:RemoveFromContinueWatching

I didn't draft a PR yet since this design hasn't been signed off of yet.

**Reply by cvium** [reply]
I guess I misunderstood and got a little ahead of myself. In my mind Next Up and Continue Watching are already merged into one. And in that scenario I would prefer it being more static.

In the current version of CW, I guess nathangur's implementation is OK. I do worry about the impact on the already-quite-poor Next Up query though. I'm also worried that people will remove a Series and forget and complain about CW/Next Up not working.

**Reply by nathangur** [reply]
> I guess I misunderstood and got a little ahead of myself. In my mind Next Up and Continue Watching are already merged into one. And in that scenario I would prefer it being more static.
> 
> In the current version of CW, I guess nathangur's implementation is OK. I do worry about the impact on the already-quite-poor Next Up query though. I'm also worried that people will remove a Series and forget and complain about CW/Next Up not working.

@cvium Could you explain what you mean by CW and NU being merged? I'm not sure I'm following what you are trying to say. 

As for someone forgetting that they removed an item or series there's not much we can do about that either way.

**Reply by cvium** [reply]
Having Next Up and Continue Watching as separate sections is inefficient use of UI real estate.

There's plenty of ways to make it more visible which series/whatever have been hidden

**Reply by nathangur** [reply]
@cvium I agree with that sentiment, however I don't see the full redesign for the Jellyfin clients being fully realized for quite some time. And since this change isn't too obtrusive I don't agree with not merging the proposed design or another one based on the fact that some time in the future the design or the backend of the server/clients will be overhauled. Especially since an overhaul of that scale would take years to be fully realized. To be clear this isn't a slight against the pace of Jellyfin's development it's just an unfortunate truth. Cheers :)

**Reply by Soitora** [reply]
Agreed, it's better to have this implemented sooner rather than later, a merging of the two section makes sense, but I don't feel like it's related to this feature in particular.

**Reply by cvium** [reply]
Did I say that a potential merge of CW&NU would be a blocker though? 😉 The concerns are real though.

**Reply by nathangur** [reply]
I agree with your concerns but I think that discussing those changes would be better suited [here](https://github.com/jellyfin/jellyfin-meta/discussions/136) for instance. My apologies if this sounds passive aggressive that isn't my intention

**Reply by barrelltitor** [reply]
Would be nice to just get the feature added already and not worry about adjacent issues. Jellyfin moves at a snails pace in general, more artificial waiting isn't good

**Reply by cvium** [reply]
> I agree with your concerns but I think that discussing those changes would be better suited [here](https://github.com/jellyfin/jellyfin-meta/discussions/136) for instance. My apologies if this sounds passive aggressive that isn't my intention 

My concerns are with your implementation, not Ghost's design proposal.

It would be nice if you'd all stop piling on on something I never said.

**Reply by barrelltitor** [reply]
I didn't mean to add to the pile, its just that jf as a project moves very slowly. We've had 2 years of just CFcore. If the solution is not perfect but we never have a better one might as well go for the worse solution for now

**Reply by BrennanWoodbury** [reply]
Is it possible that we're letting perfect be the enemy of good? I don't mean to say we should push garbage, but this has been one of the most requested features for quite some time, and there's been no meaningful movement on this. 

Nearly 3 years ago now, @OccamsChainsaw42 and I responded to this [blog post](https://jellyfin.org/posts/a-call-for-developers) by trying to do an initial implementation for this feature alongside @nathangur. It was ultimately shot down due to the implementation being too complicated, and the pending EF core migration. Migration is over and done with now, are we really going to continue to postpone this even further? There are people willing and able to do the work, but an imperfect implementation is yet again blocking progress to software that better serves the end user. 

Again, this is one of the most requested features for Jellyfin to date. 

This is not meant to be a personal attack on the maintainers, I have no doubt you all work very hard at a job that is almost entirely thankless -- but there seems to be a major disconnect here and something needs to change.

**Reply by cvium** [reply]
> I do worry about the impact on the already-quite-poor Next Up query though. I'm also worried that people will remove a Series and forget and complain about CW/Next Up not working.

Instead of bemoaning the pace and accusing me of wanting a perfect implementation, you could respond to this in a non-handwavy way.

**Reply by BrennanWoodbury** [reply]
Fair enough --

On the Next Up query: the filter goes in `GetNextUpSeriesKeys`. That call runs with no limit and a `MinValue` date cutoff, so it already hands every series you've ever watched to `GetNextUpEpisodesBatch`, which materializes their episodes and loops per series. Excluding a series there means its episodes never enter that batch at all — it shrinks the input rather than adding another filter on top. And it's guarded (`if (ids.Count > 0)`), so anyone who's hidden nothing gets exactly the query they get today.

On people forgetting: playing anything in a hidden series clears it. Then secondarily, a hidden-items list in settings, maybe linked to from a context menu in the CW list? — same pattern as `LatestItemExcludes`. Curious what y'all think on the hidden item list and how and if to implement.  

If this is good enough, is the preferred path implementing `ExcludedFromResume` from #136 , or reshaping @nathangur's branch to match it?

Otherwise, of course, open to discussion.

**Comment by goggybox** [comment]
Hi all, I wanted to return to this topic. I think @jcreek's proposal is reasonable, especially now that the rewrite to EFCore is done. I was hoping to get some feedback from @JPVenson or @thornbill on this approach, and whether you'd still prefer to hold off until more of the existing code is overhauled? This feature is pretty important to me personally (it's one of the reasons I'm still hesitant to fully migrate to Plex), so I'm happy to contribute wherever I can if you are open to draft PRs for this feature.

**Reply by JPVenson** [reply]
just as an FYI, we are currently pretty busy with the 12.0 release so feedback on this might not happen in the short term as its pretty involved

**Comment by jcreek** [comment]
I'm also still very happy to work on this 😄

**Comment by Soitora** [comment]
As a recent Plex migrant, this not being a feature already is honestly the one thing that keeps me from loving Jellyfin outright, happy to see that other people seem to agree. Now hopefully it can be closer to reality after v12 releases. 😊 

The table posted by jcreek looks about right, it's an optimal way to handle the flags or db entries

**Comment by Shadowghost** [comment]
Just as a headsup #136 also includes a full implementation proposal for this feature alongside the rewritten watched tracking

**Reply by nathangur** [reply]
I see. I provided a implementation for this specific feature for CW and NU here: https://github.com/jellyfin/jellyfin-meta/discussions/50#discussioncomment-17945211

I don't know if this implementation would fall in nicely with that proposed redesign so please take a look. Cheers

**Comment by jazzathedev** [comment]
This seriously needs to exist, I know design paralysis is real but someone in the core decision making team needs to step up and let a willing developer implement it, the jank required to achieve this through plugins is unusable (not to mention plugins themselves being hell to work with as an end user)
